File: cpp/api/src/base/Waiting.cpp
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

#include "Waiting.h"
#include "WaitingSet.h"
#include "Application.h"
#include "This.h"
#include <iostream>

using namespace std;

namespace cameo {

Waiting::Waiting(Waiting::Function function)
: m_function{function} {

	// Add the object in the waiting set if This exists.
	if (This::m_instance.m_inited) {
		This::m_instance.m_waitingSet->add(this);
	}
}

Waiting::~Waiting() {

	// Remove the object in the waiting set if This exists.
	if (This::m_instance.m_inited) {
		This::m_instance.m_waitingSet->remove(this);
	}
}

void Waiting::cancel() {
	m_function();
}

}

File: java/api/src/main/java/eu/ill/cameo/api/base/Waiting.java
/*
 * CAMEO
 *
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under BSD 3-Clause and GPL-v3 as described in license files.
 * You may not use this work except in compliance with the Licences.
 *
 */

package eu.ill.cameo.api.base;

/**
 * Base class for the waiting classes.
 * The class enables to cancel waitings and terminate the execution (used by This.terminate). 
 */
public abstract class Waiting {

	/**
	 * Adds this object to the set of waitings.
	 */
	public void add() {
		// Add the waiting if This exists.
		if (This.instance != null) {
			This.instance.getWaitingSet().add(this);
		}	
	}
	
	/**
	 * Removes this object from the set of waitings.
	 */
	public void remove() {
		// Remove the waiting if This exists.
		if (This.instance != null) {
			This.instance.getWaitingSet().remove(this);
		}
	}
	
	/**
	 * Cancels the waiting.
	 */
	public abstract void cancel();
	
	/**
	 * Terminates the waiting object.
	 */
	public abstract void terminate();
}